File: javascript/word-count/word-count.js
var Words = function(){};

Words.prototype.count = function(sentence){
  var wordsArray = sentence.trim().split(/\s+/g);
  var wordHistogram = {};

  for (i in wordsArray){
    var word = wordsArray[i].toLowerCase();

    if (wordHistogram.hasOwnProperty(word)){
      wordHistogram[word]++;
    } else {
      wordHistogram[word] = 1;
    }
  }

  return wordHistogram;
};

module.exports = Words;

File: elixir/word-count/word_count.exs
defmodule Words do
  @doc """
  Count the number of words in the sentence.
  Words are compared case-insensitively.
  """
  @spec count(String.t) :: map
  def count(sentence) do
    sentence
    |> String.downcase()
    |> String.split([",", ":", "@", "&", "^", "%", "!", "?", "$", "_", " "], trim: true)
    |> make_hist()
  end

  def make_hist(word_list) do
    word_list
    |> Enum.reduce(%{}, fn (word, acc) -> Map.update(acc, word, 1, &(&1 + 1)) end)
  end
end
